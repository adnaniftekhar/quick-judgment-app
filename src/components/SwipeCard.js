import React from 'react';
import { Animated, PanResponder, Dimensions, StyleSheet } from 'react-native';
import { Card } from './Card';

const SCREEN_WIDTH = Dimensions.get('window').width;
const SWIPE_THRESHOLD = 0.25 * SCREEN_WIDTH;
const SWIPE_OUT_DURATION = 250;

export const SwipeCard = ({ data, onSwipeLeft, onSwipeRight }) => {
  const position = new Animated.ValueXY();

  const panResponder = PanResponder.create({
    onStartShouldSetPanResponder: () => true,
    onPanResponderMove: (_, gesture) => {
      position.setValue({ x: gesture.dx, y: gesture.dy });
    },
    onPanResponderRelease: (_, gesture) => {
      if (gesture.dx > SWIPE_THRESHOLD) {
        forceSwipe('right');
      } else if (gesture.dx < -SWIPE_THRESHOLD) {
        forceSwipe('left');
      } else {
        resetPosition();
      }
    }
  });

  const forceSwipe = (direction) => {
    const x = direction === 'right' ? SCREEN_WIDTH : -SCREEN_WIDTH;
    Animated.timing(position, {
      toValue: { x, y: 0 },
      duration: SWIPE_OUT_DURATION,
      useNativeDriver: false
    }).start(() => onSwipeComplete(direction));
  };

  const onSwipeComplete = (direction) => {
    direction === 'right' ? onSwipeRight() : onSwipeLeft();
    position.setValue({ x: 0, y: 0 });
  };

  const resetPosition = () => {
    Animated.spring(position, {
      toValue: { x: 0, y: 0 },
      useNativeDriver: false
    }).start();
  };

  const getCardStyle = () => {
    const rotate = position.x.interpolate({
      inputRange: [-SCREEN_WIDTH * 1.5, 0, SCREEN_WIDTH * 1.5],
      outputRange: ['-120deg', '0deg', '120deg']
    });

    return {
      ...position.getLayout(),
      transform: [{ rotate }]
    };
  };

  return (
    <Animated.View
      style={[styles.container, getCardStyle()]}
      {...panResponder.panHandlers}
    >
      <Card data={data} />
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    width: '100%',
    height: '60%',
    alignItems: 'center',
    justifyContent: 'center'
  }
});
