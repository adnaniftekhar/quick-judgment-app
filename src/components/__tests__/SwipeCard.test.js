import React from 'react';
import renderer from 'react-test-renderer';

// Mock the Card component first
jest.mock('../Card', () => ({
  Card: () => 'Card'
}));

// Create a mock for setValue that we can track
const setValueMock = jest.fn();

// Enhanced mock for react-native
jest.mock('react-native', () => {
  const animationCallback = jest.fn();
  const mockSetValue = jest.fn();  // Define mock inside factory function
  
  return {
    View: 'View',
    Text: 'Text',
    Animated: {
      View: 'View',
      ValueXY: jest.fn(() => ({
        x: { interpolate: jest.fn(() => 'rotate') },
        y: 0,
        getLayout: () => ({}),
        setValue: mockSetValue  // Use locally defined mock
      })),
      timing: jest.fn(() => ({
        start: (cb) => {
          animationCallback();
          cb && cb();
        }
      })),
      spring: jest.fn(() => ({ start: jest.fn() }))
    },
    PanResponder: {
      create: jest.fn(() => ({
        panHandlers: {}
      }))
    },
    Dimensions: {
      get: () => ({
        width: 375,
        height: 667
      })
    },
    StyleSheet: {
      create: styles => styles
    },
    _animationCallback: animationCallback,
    _mockSetValue: mockSetValue  // Expose for tests
  };
});

import { SwipeCard } from '../SwipeCard';

describe('SwipeCard Component', () => {
  const mockProps = {
    data: { statement: 'Test Statement' },
    onSwipeLeft: jest.fn(),
    onSwipeRight: jest.fn()
  };

  let panResponderConfig;

  beforeEach(() => {
    mockProps.onSwipeLeft.mockClear();
    mockProps.onSwipeRight.mockClear();
    require('react-native')._animationCallback.mockClear();
    
    require('react-native').PanResponder.create.mockImplementation(config => {
      panResponderConfig = config;
      return { panHandlers: {} };
    });
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <SwipeCard {...mockProps} />
    ).toJSON();
    
    expect(tree).toBeDefined();
    expect(tree.type).toBe('View');
  });

  it('renders with correct data', () => {
    const tree = renderer.create(
      <SwipeCard {...mockProps} />
    ).toJSON();
    
    expect(tree.children).toBeDefined();
  });

  it('applies container styles', () => {
    const tree = renderer.create(
      <SwipeCard {...mockProps} />
    ).toJSON();
    
    expect(tree.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          position: 'absolute',
          width: 375,
          height: '100%',
          padding: 10
        })
      ])
    );
  });

  it('handles right swipe correctly', () => {
    renderer.create(<SwipeCard {...mockProps} />);
    
    // Simulate a right swipe
    panResponderConfig.onPanResponderRelease(null, {
      dx: 200  // More than SWIPE_THRESHOLD (which is 93.75 based on our mocked width)
    });
    
    // Wait for animation to complete
    expect(require('react-native')._animationCallback).toHaveBeenCalled();
    expect(mockProps.onSwipeRight).toHaveBeenCalled();
    expect(mockProps.onSwipeLeft).not.toHaveBeenCalled();
  });

  it('handles left swipe correctly', () => {
    renderer.create(<SwipeCard {...mockProps} />);
    
    // Simulate a left swipe
    panResponderConfig.onPanResponderRelease(null, {
      dx: -200  // Less than -SWIPE_THRESHOLD
    });
    
    // Wait for animation to complete
    expect(require('react-native')._animationCallback).toHaveBeenCalled();
    expect(mockProps.onSwipeLeft).toHaveBeenCalled();
    expect(mockProps.onSwipeRight).not.toHaveBeenCalled();
  });

  it('resets position on small swipe', () => {
    renderer.create(<SwipeCard {...mockProps} />);
    
    // Simulate a small swipe that shouldn't trigger left/right
    panResponderConfig.onPanResponderRelease(null, {
      dx: 50  // Less than SWIPE_THRESHOLD
    });
    
    expect(mockProps.onSwipeLeft).not.toHaveBeenCalled();
    expect(mockProps.onSwipeRight).not.toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const tree = renderer.create(
      <SwipeCard 
        data={{ statement: 'Snapshot test statement' }}
        onSwipeLeft={() => {}}
        onSwipeRight={() => {}}
      />
    ).toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('updates position during movement', () => {
    renderer.create(<SwipeCard {...mockProps} />);
    
    // Simulate pan movement
    panResponderConfig.onPanResponderMove(null, {
      dx: 100,
      dy: 50
    });
    
    // Get the mock from the mocked module
    const { _mockSetValue } = require('react-native');
    expect(_mockSetValue).toHaveBeenCalledWith({
      x: 100,
      y: 50
    });
  });

  it('starts pan responder on touch', () => {
    renderer.create(<SwipeCard {...mockProps} />);
    
    // Verify that onStartShouldSetPanResponder returns true
    expect(panResponderConfig.onStartShouldSetPanResponder()).toBe(true);
  });
}); 