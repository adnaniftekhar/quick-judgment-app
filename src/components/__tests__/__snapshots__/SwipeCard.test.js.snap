// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeCard Component matches snapshot 1`] = `
<View
  style={
    [
      {
        "height": "100%",
        "padding": 10,
        "position": "absolute",
        "width": 375,
      },
      {
        "transform": [
          {
            "rotate": "rotate",
          },
        ],
      },
    ]
  }
>
  Card
</View>
`;
